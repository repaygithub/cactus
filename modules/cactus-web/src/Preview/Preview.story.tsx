import React from 'react'

import { Preview } from '../'

export default {
  title: 'Preview',
  component: Preview,
  parameters: { controls: { disable: true } },
} as const

const IMAGES = [
  'https://repay-merchant-resources.s3.amazonaws.com/staging/24bd1970-a677-4ca7-a4d2-e328ddd4691b/repay_logo_new.jpg',
  'http://placekitten.com/400/450',
  'http://placekitten.com/450/250',
  'http://placekitten.com/250',
  'http://placekitten.com/600/300',
  'http://placekitten.com/1250/700',
]

export const BasicUsage = (): React.ReactElement => <Preview images={IMAGES} />

export const FromChildren = (): React.ReactElement => (
  <Preview>
    {IMAGES.map((src, ix) => (
      <img src={src} key={ix} alt={ix === 0 ? 'Repay Logo' : `Cute kitten number ${ix}`} />
    ))}
  </Preview>
)

FromChildren.parameters = { storyshots: false }
