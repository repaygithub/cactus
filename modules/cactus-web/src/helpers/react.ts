import { assign, noop } from 'lodash'
import React, { SetStateAction } from 'react'

export type CloneFunc = (
  e: React.ReactElement,
  p?: Record<string, any>,
  ix?: number
) => React.ReactElement

// This can also be used just to flatten the children, but cloning seems the more likely use case.
export function cloneAll(
  children: React.ReactNode,
  cloneProps?: Record<string, any>,
  makeClone: CloneFunc = React.cloneElement
): React.ReactChild[] {
  const hasKeyPrefix = cloneProps && (cloneProps.key || cloneProps.key === 0)
  const childArray = React.Children.toArray(children) as React.ReactChild[]
  return childArray.reduce((returnVal, child, index) => {
    if (!child || typeof child === 'string' || typeof child === 'number') {
      returnVal.push(child)
    } else {
      const key = hasKeyPrefix ? `${cloneProps?.key}${child.key}` : child.key
      let props: Record<string, any> | undefined
      if (cloneProps) {
        props = { ...cloneProps, key }
      } else if (hasKeyPrefix) {
        props = { key }
      }
      if (child.type === React.Fragment) {
        if (child.props.children) {
          returnVal.push(...cloneAll(child.props.children, props, makeClone))
        }
      } else {
        returnVal.push(makeClone(child, props, index))
      }
    }
    return returnVal
  }, [] as React.ReactChild[])
}

type Ref<T> = React.RefCallback<T> | React.MutableRefObject<T> | null | undefined
type HybridRef<T> = React.RefCallback<T> & React.RefObject<T>
type MutHybridRef<T> = React.RefCallback<T> & React.MutableRefObject<T>

// Keep in mind that unlike a normal ref, hybrid refs are not guaranteed to be stable.
export function useMergedRefs<T>(...refs: Ref<T>[]): HybridRef<T> {
  const refFunc: any = (value: T | null) => {
    if (!value) return
    hybridRef.current = value
    for (const ref of refs) {
      if (!ref) {
        continue
      } else if (typeof ref === 'function') {
        ref(value)
      } else {
        ref.current = value
      }
    }
  }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const hybridRef = React.useCallback<MutHybridRef<T>>(refFunc, refs)
  if (hybridRef === refFunc) {
    Object.defineProperty(hybridRef, 'current', { writable: true, enumerable: true })
    Object.preventExtensions(hybridRef)
  }
  return hybridRef
}

type Func<R = any> = (...args: any[]) => R
type NotFunc<T> = T extends Func ? never : T

interface ValueHook {
  <T extends Func>(mkValue: T, args: Parameters<T>): ReturnType<T>
  <T extends Func, D = Parameters<T>>(
    mkValue: T,
    deps: D,
    compareDependencies: (prev: Parameters<T>, current: D) => Parameters<T>
  ): ReturnType<T>
  <T>(value: NotFunc<T>, deps: unknown[]): T
  <T, D>(value: NotFunc<T>, deps: D, compareDependencies: (prev: D, current: D) => D): T
}

interface Value<T> {
  value: T
  dependencies: unknown[]
}

const shallowCmp = (prev: any[], current: any[]): any[] => {
  if (prev.length !== current.length || !prev.every((d, i) => d === current[i])) {
    return current
  }
  return prev
}

// Similar to useMemo/useCallback except stability is guaranteed.
export const useValue: ValueHook = <T>(val: any, deps: any[], cmp = shallowCmp): T => {
  const ref = React.useRef(SENTINEL as Value<T>)
  let value = ref.current.value
  const prevDeps = ref.current.dependencies
  const dependencies = cmp(prevDeps, deps)
  if (dependencies !== prevDeps) {
    value = typeof val === 'function' ? val(...dependencies) : val
    ref.current = { value, dependencies }
  }
  return value
}
const SENTINEL: unknown = { dependencies: [Object.create(null)] }

interface UseBox {
  <T>(box: T): T
  // Mutable box + immutable initialization
  <M, I extends Func>(box: M, init: I, ...args: Parameters<I>): M & ReturnType<I>
}
type Obj = Record<string, any>

// An immutable container with mutable contents; basically, a self-updating ref.
export const useBox: UseBox = (box: Obj, init?: Func<Obj>, ...args: any[]): Obj => {
  const ref = React.useRef<Obj>()
  if (ref.current === undefined) {
    ref.current = init ? init(...args) : {}
  }
  return assign(ref.current, box)
}

type StateSetterWithCallback<S> = (v: SetStateAction<S>, callback?: () => void) => void

export function useStateWithCallback<S>(
  initialState: S | (() => S)
): [S, StateSetterWithCallback<S>] {
  const [state, setState] = React.useState<S>(initialState)
  const storedCallback = React.useRef<() => void>(noop)

  const setStateWithCallback = React.useCallback<StateSetterWithCallback<S>>(
    (v, callback = noop) => {
      storedCallback.current = callback
      setState(v)
    },
    []
  )

  React.useEffect(() => {
    storedCallback.current()
    storedCallback.current = noop
  }, [state])

  return [state, setStateWithCallback]
}

const inc = (x: number) => x + 1

export const useRenderTrigger = (): (() => void) => React.useReducer(inc, 0)[1]

type SemiControlProps<T extends React.ElementType> = {
  input: T
  value: unknown
  onChange: React.ChangeEventHandler<unknown>
} & Omit<React.ComponentProps<T>, 'as' | 'ref' | 'value' | 'onChange'>

interface ControlBox {
  value: unknown
  isFocused: boolean
  onFocus?: React.FocusEventHandler<unknown>
  handleFocus: React.FocusEventHandler<unknown>
  onBlur?: React.FocusEventHandler<unknown>
  handleBlur: React.FocusEventHandler<unknown>
  onChange: React.ChangeEventHandler<unknown>
  handleChange: React.ChangeEventHandler<unknown>
}

const initControlBox = (trigger: () => void): ControlBox => {
  const vars: ControlBox = {
    value: null,
    isFocused: false,
    onChange: noop,
    handleFocus: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = true
      vars.onFocus?.(e)
    },
    handleBlur: (e: React.FocusEvent<unknown>) => {
      vars.isFocused = false
      vars.onBlur?.(e)
      trigger() // Now rerender with the parent's value.
    },
    handleChange: (e: React.ChangeEvent<{ value: unknown }>) => {
      vars.value = e.target.value
      vars.onChange(e)
      trigger()
    },
  }
  return vars
}

// Use this to make an input "uncontrolled" while focused, but controlled otherwise;
// this is useful when valid values are limited but you don't want to interfere
// with user input by trying to normalize the value as they type.
export const SemiControlled = <T extends React.ElementType>(
  props: SemiControlProps<T>
): React.ReactElement => {
  const { input, value: controlledValue, onFocus, onBlur, onChange, ...rest } = props
  const trigger = useRenderTrigger()
  const vars = useBox({ onFocus, onBlur, onChange, controlledValue }, initControlBox, trigger)
  if (!vars.isFocused) {
    vars.value = controlledValue
  }
  return React.createElement(input, {
    ...rest,
    onFocus: vars.handleFocus,
    onBlur: vars.handleBlur,
    onChange: vars.handleChange,
    value: vars.value,
  })
}
