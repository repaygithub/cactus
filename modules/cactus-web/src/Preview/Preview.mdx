---
name: Preview
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import { LiveProvider, LiveEditor, LiveError, LivePreview } from 'react-live'
import Preview from './Preview'

# Preview

### Try It Out

export const code = `
  <Preview>
    <img
      src="https://repay-merchant-resources.s3.amazonaws.com/staging/24bd1970-a677-4ca7-a4d2-e328ddd4691b/repay_logo_new.jpg"
      alt="Repay Logo"
    />
    <img src="http://placekitten.com/400/450" alt="Cute kitten number 1" />
    <img src="http://placekitten.com/450/250" alt="Cute kitten number 2" />
    <img src="http://placekitten.com/600/300" alt="Cute kitten number 3" />
  </Preview>
`

<LiveProvider code={code} scope={{ Preview }}>
  <LiveEditor style={livePreviewStyle} />
  <LiveError />
  <LivePreview />
</LiveProvider>

## Best practices

- There are two ways the `Preview` component can be used.
  1. Passing the image sources through the `images` prop:
      - This method should only be used if accessibility is not a concern.
  2. Rendering images as children:
      - This gives you the ability to set some `alt` text on the images to make them accessible.
- Another important note on accessibility is that the component exposes a `phrases` prop which you can use to set
labels for the arrow buttons and the close button.
- The `Preview` component exposes `width` and `height` props if you need to resize it. However, the minimum width is
50% and the minimum height is 200px.

## Basic usage

The examples below explain how to use the `Preview` component using the `images` prop and using children, respectively.

```jsx
const IMAGES = ['http://placekitten.com/400/450', 'http://placekitten.com/600/300']

<Preview images={IMAGES} />
```

OR

```jsx
<Preview>
  <img src="http://placekitten.com/400/450" alt="Cute kitten 1" />
  <img src="http://placekitten.com/600/300" alt="Cute kitten 2" />
</Preview>
```

## Properties

<PropsTable of={Preview} />
