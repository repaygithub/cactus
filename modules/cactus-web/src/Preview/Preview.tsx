import { NavigationChevronLeft, NavigationChevronRight, NavigationClose } from '@repay/cactus-icons'
import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'
import { height, HeightProps, margin, MarginProps, width, WidthProps } from 'styled-system'

import Dimmer from '../Dimmer/Dimmer'
import Flex from '../Flex/Flex'
import { keyDownAsClick, preventAction } from '../helpers/a11y'
import { flexItem, FlexItemProps } from '../helpers/flexItem'
import { boxShadow, radius } from '../helpers/theme'
import IconButton from '../IconButton/IconButton'

interface PreviewProps
  extends MarginProps,
    WidthProps,
    HeightProps,
    FlexItemProps,
    React.HTMLAttributes<HTMLDivElement> {
  images?: string[]
  phrases?: Phrases
}

interface Phrases {
  previous?: string
  next?: string
  close?: string
}

const DEFAULT_PHRASES: Phrases = {
  previous: 'Go to the previous image',
  next: 'Go to the next image',
  close: 'Close the image',
}

export const Preview = React.forwardRef<HTMLDivElement, PreviewProps>(
  ({ images = [], children, phrases: passedPhrases, ...rest }, ref) => {
    const [currentIndex, setCurrentIndex] = React.useState<number>(0)
    const [imageSelected, setImageSelected] = React.useState<boolean>(false)
    const selectedImageRef = React.useRef<HTMLImageElement | null>(null)
    const closeButtonRef = React.useRef<HTMLButtonElement | null>(null)
    const imgComponents = images.length
      ? images.map((src) => <img src={src} />)
      : React.Children.toArray(children)
    const multipleImages = imgComponents.length > 1
    const phrases = { ...DEFAULT_PHRASES, ...passedPhrases }

    React.useEffect(() => {
      // REACT18-COMPAT: This could theoretically cause issues when remounting,
      // but I don't think it's likely give how focusing the image takes over
      // the screen: any interaction likely to "hide" the component will also
      // likely close the selected image.
      if (imageSelected) {
        closeButtonRef.current?.focus()
      }

      const handleBodyClick = (event: MouseEvent): void => {
        const { target } = event
        if (
          target instanceof Node &&
          imageSelected &&
          !selectedImageRef?.current?.contains(target)
        ) {
          setImageSelected(false)
        }
      }

      const handleBodyKeyDown = (event: KeyboardEvent) => {
        if (imageSelected && event.key === 'Escape') {
          setImageSelected(false)
        }
      }

      document.body.addEventListener('click', handleBodyClick)
      document.body.addEventListener('keydown', handleBodyKeyDown)

      return () => {
        document.body.removeEventListener('click', handleBodyClick)
        document.body.removeEventListener('keydown', handleBodyKeyDown)
      }
    }, [imageSelected])

    const handleLeftArrowClick = () => {
      setCurrentIndex((index) => index - 1)
    }

    const handleRightArrowClick = () => {
      setCurrentIndex((index) => index + 1)
    }

    const handleImageClick = () => {
      setImageSelected(true)
    }

    return (
      <PreviewBox ref={ref} justify={multipleImages ? 'space-between' : 'center'} {...rest}>
        {multipleImages && (
          <IconButton
            onClick={handleLeftArrowClick}
            onKeyDown={keyDownAsClick}
            onKeyUp={preventAction}
            disabled={currentIndex === 0}
            label={phrases.previous}
          >
            <NavigationChevronLeft aria-hidden="true" />
          </IconButton>
        )}
        {React.cloneElement(imgComponents[currentIndex] as JSX.Element, {
          onClick: handleImageClick,
          onKeyDown: keyDownAsClick,
          onKeyUp: preventAction,
          tabIndex: 0,
        })}
        {multipleImages && (
          <IconButton
            onClick={handleRightArrowClick}
            onKeyDown={keyDownAsClick}
            onKeyUp={preventAction}
            disabled={currentIndex === imgComponents.length - 1}
            label={phrases.next}
          >
            <NavigationChevronRight aria-hidden="true" />
          </IconButton>
        )}
        {imageSelected && (
          <Dimmer active>
            <Flex height="100%" flexDirection="column" alignItems="center" justifyContent="center">
              <ScrollWrapper>
                <Flex width="100%" mb={4}>
                  <StyledIconButton
                    ref={closeButtonRef}
                    ml="auto"
                    onKeyDown={keyDownAsClick}
                    onKeyUp={preventAction}
                    label={phrases.close}
                    variant="action"
                  >
                    <NavigationClose aria-hidden="true" />
                  </StyledIconButton>
                </Flex>
                <ImageBackground>
                  {React.cloneElement(imgComponents[currentIndex] as JSX.Element, {
                    ref: selectedImageRef,
                    tabIndex: 0,
                  })}
                </ImageBackground>
              </ScrollWrapper>
            </Flex>
          </Dimmer>
        )}
      </PreviewBox>
    )
  }
)

const ImageBackground = styled.div`
  display: flex;
  background-color: ${(p) => p.theme.colors.white};
`

const PreviewBox = styled.div<{ justify: 'space-between' | 'center' }>`
  box-sizing: border-box;
  padding-left: ${(p) => p.theme.space[7]}px;
  padding-right: ${(p) => p.theme.space[7]}px;
  background-color: ${(p) => p.theme.colors.lightContrast};
  border-radius: ${radius(8)};
  min-height: 200px;
  min-width: 50%;
  width: 50%;
  height: 440px;
  ${width}
  ${height}
  display: flex;
  justify-content: ${(p) => p.justify};
  align-items: center;
  ${margin}
  ${flexItem}

  img {
    display: inline;
    max-width: 80%;
    max-height: 80%;
    flex-shrink: 0;
    cursor: pointer;
    border-radius: ${radius(8)};
    ${(p) => boxShadow(p.theme, 2)}
  }

  ${ImageBackground} img {
    max-width: 100%;
    max-height: 100%;
    border-radius: 0;
    box-shadow: none;
    width: 100%;
    height: 100%;
    flex-shrink: 0;
  }
`

const ScrollWrapper = styled.div`
  max-width: 80%;
  max-height: 100%;
  overflow: auto;
  padding: ${(p) => p.theme.space[3]}px;
`

const StyledIconButton = styled(IconButton)`
  color: ${(p) => p.theme.colors.white};

  &:hover {
    color: ${(p) => p.theme.colors.white};
  }
`

Preview.propTypes = {
  images: PropTypes.arrayOf(PropTypes.string.isRequired),
  // @ts-ignore
  phrases: PropTypes.shape({
    previous: PropTypes.string,
    next: PropTypes.string,
    close: PropTypes.string,
  }),
}
Preview.displayName = 'Preview'

export default Preview
