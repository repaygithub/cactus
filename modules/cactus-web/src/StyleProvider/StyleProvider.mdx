---
name: StyleProvider
menu: Components
---

import PropsTable from 'website-src/components/PropsTable'
import { livePreviewStyle } from '../helpers/constants'
import StyleProvider from './StyleProvider'
import cactusTheme from '@repay/cactus-theme'

# StyleProvider

The `StyleProvider` is a wrapper for `styled-components`' `ThemeProvider` component which gives you the option to include some global styles.

# Best practices

- Make sure that any components from `@repay/cactus-web` are children of `StyleProvider`. This will ensure that each component has the correct styles.

## Basic usage

```jsx
import React from 'react'
import { StyleProvider } from '@repay/cactus-web'
import cactusTheme from '@repay/cactus-theme'

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      // Include routes/rest of application here
    </StyleProvider>
  )
}
```

# Breakpoints

The `StyleProvider` has a media queries scale which is based on the provided breaking points and is added to the given theme. The media queries have screen sizes small, medium, large, and extraLarge. The set media queries will be larger than the default, for a mobile first approach.

Props can also vary depending on the current media query. This is done by setting a prop to an array, where the array holds different values for each media query size. The first element of the array maps to the default screen size, and each element after maps to the set media queries from smallest to largest. This is used below in the example code to set the background color.

For more information on Breakpoints and Media Queries visit: https://styled-system.com/theme-specification#breakpoints

## Media Queries Usage

```jsx
import React from 'react'
import { StyleProvider, Box } from '.@repay/cactus-web'
import styled from 'styled-components'

const BreakpointBox = styled(Box)`
  border-radius: 15%;

  ${p => p.theme.mediaQueries.small} {
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extraLarge} {
    border-radius: 50%;
  }
`
const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const App = () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <BreakpointBox backgroundColor={backgroundColors} width="100px" height="100px" />
    </StyleProvider>
  )
}
```

# Properties

<PropsTable of={StyleProvider} staticProp="StyleProvider" />
