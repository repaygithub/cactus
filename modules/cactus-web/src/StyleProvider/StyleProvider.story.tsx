import React from 'react'

import { storiesOf } from '@storybook/react'
import Box from '../Box/Box'
import cactusTheme from '@repay/cactus-theme'
import Grid from '../Grid/Grid'
import styled from 'styled-components'
import StyleProvider from './StyleProvider'

const backgroundColors = ['springgreen', 'mediumturquoise', 'cornflowerblue', 'thistle', 'pink']

const BreakpointBox = styled(Box)`
  max-width: 320px;
  width: 50px;
  height: 50px;
  border-radius: 15%

  ${p => p.theme.mediaQueries.small} {
    max-width: 708px;
    width: 75px;
    height 75px;
    border-radius: 25%;
  }
  ${p => p.theme.mediaQueries.medium} {
    max-width: 964px;
    width: 100px;
    height: 100px;
    border-radius: 35%;
  }
  ${p => p.theme.mediaQueries.large} {
    max-width: 1140px;
    width: 150px;
    height: 150px;
    border-radius: 45%;
  }
  ${p => p.theme.mediaQueries.extralarge} {
    max-width: 1380px;
    width: 200px;
    height: 200px;
    border-radius: 50%;
  }
`

storiesOf('Style Provider', module).add('Component Adjusments based on Media Queries', () => {
  return (
    <StyleProvider theme={cactusTheme} global={true}>
      <Grid justify="center">
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={4}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
        <Grid.Item tiny={2}>
          <BreakpointBox backgroundColor={backgroundColors} />
        </Grid.Item>
      </Grid>
    </StyleProvider>
  )
})
