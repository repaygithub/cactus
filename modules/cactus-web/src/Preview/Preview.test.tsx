import { fireEvent, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as React from 'react'

import { StyleProvider } from '../StyleProvider/StyleProvider'
import Preview from './Preview'

const IMAGES = ['http://placekitten.com/400/450', 'http://placekitten.com/450/250']

describe('component: Preview', () => {
  describe('Mouse Interactions', () => {
    test('Should be able to change the image using the arrows', () => {
      const { getByAltText, queryByAltText, getByLabelText } = render(
        <StyleProvider>
          <Preview>
            {IMAGES.map((src, ix) => (
              <img src={src} alt={`Cute kitten number ${ix + 1}`} key={ix} />
            ))}
          </Preview>
        </StyleProvider>
      )

      const leftArrow = getByLabelText('Go to the previous image')
      const rightArrow = getByLabelText('Go to the next image')

      expect(getByAltText('Cute kitten number 1')).toBeInTheDocument()
      expect(queryByAltText('Cute kitten number 2')).not.toBeInTheDocument()
      expect(leftArrow).toBeDisabled()
      userEvent.click(rightArrow)
      expect(getByAltText('Cute kitten number 2')).toBeInTheDocument()
      expect(queryByAltText('Cute kitten number 1')).not.toBeInTheDocument()
      expect(rightArrow).toBeDisabled()
      userEvent.click(leftArrow)
      expect(getByAltText('Cute kitten number 1')).toBeInTheDocument()
    })

    test('should be able to expand an image', () => {
      const { getByAltText, getAllByAltText, getByLabelText } = render(
        <StyleProvider>
          <Preview>
            {IMAGES.map((src, ix) => (
              <img src={src} alt={`Cute kitten number ${ix + 1}`} key={ix} />
            ))}
          </Preview>
        </StyleProvider>
      )

      const image = getByAltText('Cute kitten number 1')
      userEvent.click(image)
      const closeButton = getByLabelText('Close the image')
      expect(document.activeElement).toBe(closeButton)
      expect(getAllByAltText('Cute kitten number 1')[1]).toBeInTheDocument()
      userEvent.click(closeButton)
      expect(getAllByAltText('Cute kitten number 1').length).toBe(1)
    })
  })

  describe('Keyboard Interactions', () => {
    test('Should be able to change the image using the arrows', () => {
      const { getByAltText, queryByAltText, getByLabelText } = render(
        <StyleProvider>
          <Preview>
            {IMAGES.map((src, ix) => (
              <img src={src} alt={`Cute kitten number ${ix + 1}`} key={ix} />
            ))}
          </Preview>
        </StyleProvider>
      )

      const leftArrow = getByLabelText('Go to the previous image')
      const rightArrow = getByLabelText('Go to the next image')

      expect(getByAltText('Cute kitten number 1')).toBeInTheDocument()
      expect(queryByAltText('Cute kitten number 2')).not.toBeInTheDocument()
      expect(leftArrow).toBeDisabled()
      fireEvent.keyDown(rightArrow, { key: 'Enter' })
      expect(getByAltText('Cute kitten number 2')).toBeInTheDocument()
      expect(queryByAltText('Cute kitten number 1')).not.toBeInTheDocument()
      expect(rightArrow).toBeDisabled()
      fireEvent.keyDown(leftArrow, { key: 'Enter' })
      expect(getByAltText('Cute kitten number 1')).toBeInTheDocument()
    })

    test('Should be able to expand an image', () => {
      const { getByAltText, getAllByAltText, getByLabelText } = render(
        <StyleProvider>
          <Preview>
            {IMAGES.map((src, ix) => (
              <img src={src} alt={`Cute kitten number ${ix + 1}`} key={ix} />
            ))}
          </Preview>
        </StyleProvider>
      )

      const image = getByAltText('Cute kitten number 1')
      fireEvent.keyDown(image, { key: 'Enter' })
      const closeButton = getByLabelText('Close the image')
      expect(document.activeElement).toBe(closeButton)
      expect(getAllByAltText('Cute kitten number 1')[1]).toBeInTheDocument()
      fireEvent.keyDown(closeButton, { key: 'Enter' })
      expect(getAllByAltText('Cute kitten number 1').length).toBe(1)
    })
  })
})
