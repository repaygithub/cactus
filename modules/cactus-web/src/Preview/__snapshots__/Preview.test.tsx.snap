// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component: Preview snapshot 1`] = `
.c2 {
  vertical-align: middle;
}

.c4 {
  vertical-align: middle;
}

.c1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 1px;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
  position: relative;
  overflow: visible;
  color: hsl(0,0%,70%);
  cursor: not-allowed;
  font-size: 24px;
}

.c1::-moz-focus-inner {
  border: 0;
}

.c1:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 8px);
  width: calc(100% + 8px);
  border: 1px solid;
  border-radius: 8px;
  border-color: hsl(200,96%,35%);
  box-sizing: border-box;
}

.c3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  box-sizing: border-box;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 1px;
  border: none;
  background: transparent;
  outline: none;
  cursor: pointer;
  position: relative;
  overflow: visible;
  color: hsl(200,96%,11%);
  font-size: 24px;
}

.c3::-moz-focus-inner {
  border: 0;
}

.c3:focus::after {
  content: '';
  display: block;
  position: absolute;
  height: calc(100% + 8px);
  width: calc(100% + 8px);
  border: 1px solid;
  border-radius: 8px;
  border-color: hsl(200,96%,35%);
  box-sizing: border-box;
}

.c3:hover {
  color: hsl(200,96%,35%);
}

.c0 {
  box-sizing: border-box;
  padding-left: 40px;
  padding-right: 40px;
  background-color: hsl(200,29%,90%);
  border-radius: 4px;
  min-height: 200px;
  min-width: 50%;
  width: 50%;
  height: 440px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 img {
  display: inline;
  max-width: 80%;
  max-height: 80%;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  cursor: pointer;
  border-radius: 4px;
  box-shadow: 0px 9px 24px hsla(200,96%,35%,0.3);
}

.c0 img[data-selected='true'] {
  width: 50%;
}

<div>
  <div
    class="c0"
  >
    <button
      aria-label="Go to the previous image"
      class="c1"
      disabled=""
      type="button"
      variant="standard"
    >
      <svg
        aria-hidden="true"
        class="c2"
        fill="currentcolor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.7071068,2.70710678 C18.0976311,2.31658249 18.0976311,1.68341751 17.7071068,1.29289322 C17.3165825,0.902368927 16.6834175,0.902368927 16.2928932,1.29289322 L6.29289322,11.2928932 C5.90236893,11.6834175 5.90236893,12.3165825 6.29289322,12.7071068 L16.2928932,22.7071068 C16.6834175,23.0976311 17.3165825,23.0976311 17.7071068,22.7071068 C18.0976311,22.3165825 18.0976311,21.6834175 17.7071068,21.2928932 L8.41421356,12 L17.7071068,2.70710678 Z"
        />
      </svg>
    </button>
    <img
      src="http://placekitten.com/400/450"
      tabindex="0"
    />
    <button
      aria-label="Go to the next image"
      class="c3"
      type="button"
      variant="standard"
    >
      <svg
        aria-hidden="true"
        class="c4"
        fill="currentcolor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17.7071068,2.70710678 C18.0976311,2.31658249 18.0976311,1.68341751 17.7071068,1.29289322 C17.3165825,0.902368927 16.6834175,0.902368927 16.2928932,1.29289322 L6.29289322,11.2928932 C5.90236893,11.6834175 5.90236893,12.3165825 6.29289322,12.7071068 L16.2928932,22.7071068 C16.6834175,23.0976311 17.3165825,23.0976311 17.7071068,22.7071068 C18.0976311,22.3165825 18.0976311,21.6834175 17.7071068,21.2928932 L8.41421356,12 L17.7071068,2.70710678 Z"
          transform="matrix(-1 0 0 1 24 0)"
        />
      </svg>
    </button>
  </div>
</div>
`;
